File: backend/node/api/src/application/use-cases/complete-run.ts
import { RunRepository } from "../../infra/repository/run";
import { RunNotFoundError } from "../../domain/errors";
import { inject } from "../../infra/DI/container";

type Input = {
	runId: string;
	result: number;
};

export class CompleteRunUseCase {
	@inject("run-repository")
	declare private readonly runRepository: RunRepository;

	async execute(input: Input): Promise<[true, undefined] | [undefined, Error]> {
		const [run, error] = await this.runRepository.getById(input.runId);
		if (error) return [undefined, error];
		if (!run) return [undefined, new RunNotFoundError()];

		const [, completeError] = run.complete(input.result);
		if (completeError) return [undefined, completeError];

		const [, saveError] = await this.runRepository.update(run);
		if (saveError) return [undefined, saveError];

		return [true, undefined];
	}
}

File: backend/go/api/internal/run/use-cases/complete-run.go
package usecases

import (
	"context"

	runPkg "github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/run"
)

type CompleteRunUseCaseDependencies struct {
	RunRepository *runPkg.RunRepository
}

type CompleteRunUseCase struct {
	CompleteRunUseCaseDependencies
}

func NewCompleteRunUseCase(deps CompleteRunUseCaseDependencies) CompleteRunUseCase {
	return CompleteRunUseCase{
		CompleteRunUseCaseDependencies: deps,
	}
}

type CompleteRunInput struct {
	RunId  string
	Result float64
}

func (u *CompleteRunUseCase) Execute(input CompleteRunInput) error {
	run, err := u.RunRepository.GetById(context.Background(), input.RunId)
	if err != nil {
		return err
	}

	if run == nil {
		return runPkg.RunNotFoundError
	}

	err = run.Complete(input.Result)
	if err != nil {
		return err
	}

	err = u.RunRepository.Update(context.Background(), run)
	if err != nil {
		return err
	}

	return nil
}
